optimum/intel/openvino/modeling_decoder.py

```diff
@@ -0,0 +1,347 @@
+            "token_type_ids": None,
+        }
+
+    # Copied from transformers.models.gpt2.modeling_gpt2.GPT2LMHeadModel._reorder_cache
+    @staticmethod
+    def _reorder_cache(
+        past_key_values: Tuple[Tuple[torch.Tensor]], beam_idx: torch.Tensor
+    ) -> Tuple[Tuple[torch.Tensor]]:
+        """
+        This function is used to re-order the `past_key_values` cache if [`~PreTrainedModel.beam_search`] or
+        [`~PreTrainedModel.beam_sample`] is called. This is required to match `past_key_values` with the correct
+        beam_idx at every generation step.
+        """
+        return tuple(
+            tuple(past_state.index_select(0, beam_idx.to(past_state.device)) for past_state in layer_past)
+            for layer_past in past_key_values
+        )
+
+    def can_generate(self):
+        """Returns True to validate the check that the model using `GenerationMixin.generate()` can indeed generate."""
+        return True
+
+
+class OVDecoder:
+    def __init__(self, model: openvino.runtime.Model, device: str, ov_config: Dict, use_cache: bool):
+        self.model = model
+        self._device = device
+        self.device = torch.device("cpu")
+        self.input_names = {key.get_any_name(): idx for idx, key in enumerate(self.model.inputs)}
+        self.output_names = {key.get_any_name(): idx for idx, key in enumerate(self.model.outputs)}
+        self.key_value_input_names = [key for key in self.input_names if "key_values" in key]
+        self.key_value_output_names = [key for key in self.output_names if "present" in key]
+        self.use_cache = use_cache
+        self.num_pkv = 2
+        self.ov_config = ov_config
+        self.request = None
+
+    def forward(
+        self,
+        input_ids: torch.LongTensor,
+        attention_mask: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None,
+    ) -> CausalLMOutputWithPast:
+        self._create_inference_request()
+
+        inputs = {}
+        if past_key_values is not None:
+            # Flatten the past_key_values
+            past_key_values = tuple(
+                _contiguous_helper(np.array(past_key_value))
+                for pkv_per_layer in past_key_values
+                for past_key_value in pkv_per_layer
+            )
+            # Add the past_key_values to the decoder inputs
+            inputs = {
+                input_name: Tensor(past_key_value, shared_memory=True)
+                for input_name, past_key_value in zip(self.key_value_input_names, past_key_values)
+            }
+
+        # Create empty past_key_values for decoder_with_past first generation step
+        elif self.use_cache:
+            shape_input_ids = input_ids.shape
+            for input_name in self.key_value_input_names:
+                model_inputs = self.model.input(input_name)
+                shape = model_inputs.get_partial_shape()
+                shape[0] = shape_input_ids[0]
+                shape[2] = 0
```

Comment:
To take care about models with inverted_value_shape (e.g. bloom) 
```suggestion
                shape[2 if shape[2].is_dynamic else 1] = 0
```

Reply:
Tried your suggestion but got an error due to the batch size defintion which should change for bloom architecture, was fixed in [`06d285`](https://github.com/huggingface/optimum-intel/pull/274/commits/06d28520091e6e2e630a83db58118bc8f3cfdda1).